### ISSUE TITLE
Add NodeMonitor TPR to enable scraping kubelets

### ISSUE THREAD

**Issue by whereisaaron** [coreos/prometheus-operator#173]
[ServiceMonitor](https://github.com/coreos/prometheus-operator/blob/master/Documentation/service-monitor.md) can be used to scrape any Services (or Pods/Deployments/Daemonsets by creating  a Service to find their endpoints). This enables all the targets monitored by the [custom config](https://github.com/coreos/kube-prometheus/blob/master/assets/prometheus/prometheus.yaml) used by 'kube-prometheus' to be [configured using Prometheus Operator](https://github.com/whereisaaron/kubernetes-prometheus-operator-deployment/tree/master/prometheus-operator-config) *except for one*. AFAIK, the kubelets cannot be [discovered](https://github.com/prometheus/prometheus/blob/master/discovery/kubernetes/kubernetes.go#L204-L215) and scraped because they require a `scrape_configs` entry with `kubernetes_sd_configs: { "role": "node" }` and ServiceMonitor only supports `kubernetes_sd_configs: { "role": "endpoints" }`.

One option with be to add a 'NodeMonitor' ThirdPartyResource which has only the `jobLabel`, `interval`, `scheme`, `tlsConfig`, and `bearerToken` keys of ServiceMonitor. If such a resource is created in the same namespace as Prometheus Operator, the prometheus config would be updated to include a `scape_configs` entry something like:

```
- job_name: prometheus/kubelets/0
  scheme: https
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  kubernetes_sd_configs:
  - role: node
```

**Comment by brancz** [comment]
We decided not to go with something like this as all components are going to be self-hosted, so they are all going to be discoverable via `ServiceMonitor`s and kubelets would be the only application for the `NodeMonitor`. However, as of today self-hosted kubelets are only very experimental and not present in any of the deployment solutions, so we instead added a feature to the Prometheus Operator that allows you to synchronize kubelets into an `Endpoints` object associated with a `Service`. This feature was [released with `v0.5.0`](https://github.com/coreos/prometheus-operator/releases/tag/v0.5.0).

**Comment by galexrt** [comment]
@brancz What is the name of the service/endpoint for the kubelets?
I saw this line https://github.com/coreos/prometheus-operator/blob/0a96b0514ce20329c908395c71a0addcc974be65/cmd/operator/main.go#L48, is so the service name would be

**Comment by brancz** [comment]
The apiserver is self-hosted already :slightly_smiling_face:. The `Endpoints` object's name is configurable via the `--kubelet-object` flag in the form of `<namespace>/<name>`, so a typical one I am imagining would be `--kubelet-object=kube-system/kubelet`.

**Comment by brancz** [comment]
The general idea is that whenever a component is not self-hosted an `Endpoints` object for it can worst casn be manually be created for it.

**Comment by whereisaaron** [comment]
Great, that will work until everything is a self-hosted Service. I add the `--kubelet-object=kube-system/kubelet` and the Service is created, but no Endpoints. Checking the logs looks like there is a bug where the [Node 'name' is being used for the Endpoint 'hostname'](https://github.com/coreos/prometheus-operator/blob/dee4cbf7f86eafaefd60b8ec032855ff2e36f9b9/pkg/prometheus/operator.go#L300), but they are not compatible identifiers.

My Node names are like so (a DNS 'subdomain'): `ip-172-23-146-7.xx-region-1.compute.internal`
While Node names can be subdomain, an Endpoint `hostname` is DNS 'label' without a subdomain and so are not allow full stops. Sample log entry below (k8s 1.5.2). I assume Endpoint `hostname`s may be used to generate `kube-dns` entries or SRV records. I guess you can just keep up to the first '.'

```
ts=2017-02-26T20:30:26Z caller=operator.go:344 component=prometheusoperator msg="updating kubelet enpoints object failed"  err="Endpoints \"kubelet\" is invalid: 
[subsets[0].addresses[0].hostname: Invalid value: \"ip-172-31-148-187.xx-region-1.compute.internal\": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name$
subsets[0].addresses[1].hostname: Invalid value: \"ip-172-31-146-7.xx-region-1.compute.internal\": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name' or '123-abc'),   
subsets[0].addresses[2].hostname: Invalid value: \"ip-172-31-147-108.xx-region-1.compute.internal\": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name' or '123-abc'),   
subsets[0].addresses[3].hostname: Invalid value: \"ip-172-31-147-36.xx-region-1.compute.internal\": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name' or '123-abc'),   
subsets[0].addresses[4].hostname: Invalid value: \"ip-172-31-147-87.xx-region-1.compute.internal\": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name' or '123-abc'),   
subsets[0].addresses[5].hostname: Invalid value: \"ip-172-31-148-171.xx-region-1.compute.internal\": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name' or '123-abc')]"
```

**Comment by galexrt** [comment]
@whereisaaron See #174 I opened about this issue.

**Comment by brancz** [comment]
As the kubelet syncing feature exists and the fix for #174 was merged, can we close this @whereisaaron ?

**Comment by whereisaaron** [comment]
I think we can close it @brancz. We can do without NodeMonitor in the short term (`--kubelet-object` + #174) and the long term (self-hosted kubelet).
